docs: do it with style: interact with the bot (#2013)
sopel-irc/sopel: docs/source/plugin/bot.rst

    bot.action('does something')

Do it with style
================

.. Custom role definition to apply custom style to inline text

.. role:: red
    :class: red

.. role:: strongred
    :class: strong red

.. role:: strike
    :class: strike

.. role:: strongem
    :class: strong em

.. role:: spoiler
    :class: spoiler

> When the bot talks, replies, or acts, it can do it with style: colors,

Reviewer: ```suggestion
When the bot talks, replies, or acts, it can do so with style: colors,
```

Speaking of style, this is a stylistic suggestion.

> **bold**, *italic*, _underline_, :strike:`striked`, or ``monospace``. IRC

Reviewer: Why don't the bold and italic examples (both here and below) use those custom rST roles/CSS classes? Why are the `.em` and `.strong` classes defined if not used by these roles? Is <tt>``</tt> always guaranteed to produce monospace text in every theme? Clearly, I have questions!

Author: I use them for doing both only, which is not possible with rst.

Author: And yes, double backquote is for inline monospace/code text. No sensible person would create a Sphinx theme without using a monospace font for that. And if they do, we should change that nonetheless.

> formatting works with control code, bytes you can use to tell IRC clients how

Reviewer: ```suggestion
formatting works with control codes, bytes you can use to tell IRC clients how
```

to display some part of the text.

.. seealso::

>     If you want to know more about `IRC formatting in general`__ and some of
>     its limitations, the modern IRC documentation may be of interest to you.
> 
>     .. __: https://modern.ircdocs.horse/formatting.html

Reviewer: This just makes more sense as a link location. Alternative suggestion was to link the whole phrase, "more about IRC formatting in general".

```suggestion
    If you want to know more about IRC formatting in general and some of its
    limitations, `the modern IRC documentation`__ may be of interest to you.

    .. __: https://modern.ircdocs.horse/formatting.html
```

> However, dealing with control code yourself is not the most dev-friendly

Reviewer: ```suggestion
However, dealing with control codes yourself is not the most dev-friendly
```

approach, hence the :mod:`sopel.formatting` module. It contains various

> functions to help you create stylized texts.

Reviewer: ```suggestion
functions to help you create styled text.
```

Text styles
-----------

> Let's dive into example, starting with :func:`~sopel.formatting.bold` text::

Reviewer: ```suggestion
Let's dive into examples, starting with :func:`~sopel.formatting.bold` text::
```

    from sopel import formatting

    bot.say(formatting.bold('This is some bold text!'))

This will output a line like this:

>     [Sopel] **This is some bold text!**

Reviewer: Here I wonder how these output examples will interact with syntax highlighting. Isn't this file using Python syntax by default for pre-blocks like this?

And also: Why is Sopel's nick in the examples wrapped by `[]` instead of `<>`? The square brackets make it look (to me) like the function is secretly wrapped in `@plugin.output_prefix('[Sopel] ')`. 😁

Author: I'll change that to ` <>` .

And this is not a code/pre block, because the previous line doesn't end with `::`, but a simple, unique, `:`. In that case, it's a blockquote element.

You can use them with Python string formatting::

    emphasis = formatting.bold('important')
    bot.say('And here is the %s part.' % emphasis)

To get that kind of output:

    [Sopel] And here is the **important** part.

And you can use multiple style functions together, for example with the
:func:`~sopel.formatting.italic` function::

    word = formatting.italic('very')
    emphasis = formatting.bold('%s important' % word)
    bot.say('And here is the %s part.' % emphasis)

To get a result that looks like this:

    [Sopel] And here is the :strongem:`very` **important** part.

Colored styles
--------------

Colorized text is a bit more complicated, and Sopel tries to provide helpful
functions and constants for that: the :func:`~sopel.formatting.color` function
and the :class:`~sopel.formatting.colors` class.

The ``color`` function takes a line of text and a foreground color. It also

> accept an optional background color that uses the same color codes. The color

Reviewer: ```suggestion
accepts an optional background color that uses the same color codes. The color
```

> codes are listed by the ``colors`` class, and can be used like that::

Reviewer: ```suggestion
codes are listed by the ``colors`` class, and can be used like this::
```

    bot.say(formatting.color('Red text.', formatting.colors.RED))

> This example should produce this output:

Reviewer: ```suggestion
The above example should produce this output:
```

    [Sopel] :red:`Red text.`

You can combine colors and styles, like this::

    big = formatting.color(
        formatting.bold('WARNING'), formatting.colors.RED)
    small = formatting.italic('warning')
    bot.say('[%s] This is a %s.' % (big, small))

So you get a similar result as:

    [Sopel] [:strongred:`WARNING`] This is a *warning*.

If you want to prevent spoilers, you could be tempted to take advantage of
the background color::

    spoiler = formatting.color(
        'He was the killer.',
        formatting.colors.BLACK,
        formatting.colors.BLACK,
    )
    bot.say(spoiler)

And expect this (you need to select the text to read it):

    [Sopel] :spoiler:`He was the killer.`

> Note that not all combinations of foreground and background colors are happy
> combinations, and you should keep the usage of colors to a minimum.

Reviewer: ```suggestion
Note that not all combinations of foreground and background colors are happy
ones, and you should be mindful of using too many unnecessary colors.
```

One rephrasing option. The phrase that comes to mind is "Use sparingly", but I don't want to be quite _that_ prescriptive.

    bot.action('does something')

Do it with style
================

.. Custom role definition to apply custom style to inline text

.. role:: red
    :class: red

.. role:: boldred
    :class: bold red

.. role:: underline
    :class: underline

.. role:: strike
    :class: strike

.. role:: bolditalic
    :class: bold italic

.. role:: spoiler
    :class: spoiler

> When the bot talks, replies, or acts, it can do so with style: colors,

Reviewer: Silly idea, probably not worth implementing just because of the sheer length of markup it would require: use the UK spelling just this once and make it rainbow. 😜

Author: lol I'm sure I can do that 🌈

Author: Well, OK I can, but the time and effort is not worth it at the moment.

Reviewer: Yup. It would take seven different roles, one for each letter. Very not worth it!
